client/src/main/java/io/cloudsoft/winrm4j/client/WinRmClient.java

```diff
@@ -244,41 +252,62 @@ private int receiveCommand(String commandId, Writer out, Writer err) {
 
             //TODO use different instances of service http://cxf.apache.org/docs/developing-a-consumer.html#DevelopingaConsumer-SettingConnectionPropertieswithContexts
             setActionToContext((BindingProvider) winrm, "http://schemas.microsoft.com/wbem/wsman/1/windows/shell/Receive");
-            ReceiveResponse receiveResponse = winrm.receive(receive, RESOURCE_URI, MAX_ENVELOPER_SIZE, operationTimeout, locale, shellSelector);
-            List<StreamType> streams = receiveResponse.getStream();
-            for (StreamType s : streams) {
-                byte[] value = s.getValue();
-                if (value == null) continue;
-                if (out != null && "stdout".equals(s.getName())) {
-                    try {
-                        //TODO use passed locale?
-                        if (value.length > 0) {
-                            out.write(new String(value));
-                        }
-                        if (Boolean.TRUE.equals(s.isEnd())) {
-                            out.close();
-                        }
-                    } catch (IOException e) {
-                        throw new IllegalStateException(e);
-                    }
+
+            try {
+                numberOfReceiveCalls++;
+                ReceiveResponse receiveResponse = winrm.receive(receive, RESOURCE_URI, MAX_ENVELOPER_SIZE, operationTimeout, locale, shellSelector);
+                getStreams(receiveResponse, out, err);
+
+                System.out.println(String.format("receiveResponse: %s", receiveResponse.toString()));
+                CommandStateType state = receiveResponse.getCommandState();
+                if (COMMAND_STATE_DONE.equals(state.getState())) {
+                    return state.getExitCode().intValue();
                 }
-                if (err != null && "stderr".equals(s.getName())) {
-                    try {
-                        //TODO use passed locale?
-                        if (value.length > 0) {
-                            err.write(new String(value));
-                        }
-                        if (Boolean.TRUE.equals(s.isEnd())) {
-                            err.close();
-                        }
-                    } catch (IOException e) {
-                        throw new IllegalStateException(e);
+            } catch (SOAPFaultException e) {
+                // e.getMessage().contains("The WS-Management service cannot complete the operation within the time specified in OperationTimeout")
+                if(e.getFault().getDetail().getFirstChild().getAttributes().getNamedItem("Code").getNodeValue().equals("2150858793")) {
```

Comment:
@aledsage If some of these is null exception should be thrown again.
It is better to have the actual soap exception but I think this works as well.
I also expect that a winrs server should have error code in its response.

WDYT about catching if there is a null pointer exception and throwing the soap exception instead of the null pointer exception.